File: day7/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func readInput(path string) ([]string, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var grid []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" {
			grid = append(grid, line)
		}
	}
	return grid, scanner.Err()
}

func countSplits(grid []string) int {
	rows := len(grid)
	cols := len(grid[0])

	startCol := strings.Index(grid[0], "S")
	active := map[int]bool{startCol: true}
	splits := 0

	for r := 1; r < rows; r++ {
		next := make(map[int]bool)

		for c := range active {
			if c < 0 || c >= cols {
				continue
			}

			if grid[r][c] == '^' {
				splits++
				if c-1 >= 0 {
					next[c-1] = true
				}
				if c+1 < cols {
					next[c+1] = true
				}
			} else {
				next[c] = true
			}
		}

		if len(next) == 0 {
			break
		}
		active = next
	}

	return splits
}

func main() {
	grid, err := readInput("input.txt")
	if err != nil {
		panic(err)
	}
	fmt.Println(countSplits(grid))
}

File: day7/part1.ts
import * as fs from "fs";

function readInput(path: string): string[] {
  return fs
    .readFileSync(path, "utf-8")
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0);
}

function countSplits(grid: string[]): number {
  const rows = grid.length;
  const cols = grid[0].length;

  const startCol = grid[0].indexOf("S");
  let active = new Set<number>([startCol]);
  let splits = 0;

  for (let r = 1; r < rows; r++) {
    const next = new Set<number>();

    for (const c of active) {
      if (c < 0 || c >= cols) continue;

      if (grid[r][c] === "^") {
        splits++;
        if (c - 1 >= 0) next.add(c - 1);
        if (c + 1 < cols) next.add(c + 1);
      } else {
        next.add(c);
      }
    }

    if (next.size === 0) break;
    active = next;
  }

  return splits;
}

const grid = readInput("input.txt");
console.log(countSplits(grid));
